membership-attribute-service/app/parsers/Salesforce.scala

```diff
@@ -1,48 +1,73 @@
+    *                       the service do not process outbound messages from SF environments created by duplicating
+    *                       Prod
+    * @return
+    */
+  def parseOutboundMessage(payload: NodeSeq, organizationId: String): OutboundMessageParseError \/ Seq[OutboundMessageChange] = {
     implicit class NodeSeqOps(ns: NodeSeq) {
       def getTag(tag: String): OutboundMessageParseError \/ Node =
         (ns \ tag).headOption \/> ParsingError(s"Error while parsing the outbound message: $tag not found.\n $payload")
 
       def getText(tag: String): OutboundMessageParseError \/ String = getTag(tag).map(_.text)
     }
 
-    for {
-      orgId <- (payload \\ "notifications").getText("OrganizationId")
-      _ <- if (orgId === organizationId) ().right else OrgIdMatchingError(orgId).left
-      obj <- (payload \\ "Notification").getTag("sObject")
-      id <- obj.getText("IdentityID__c")
-    } yield {
-      val tier = (obj \ "Membership_Tier__c").map(_.text).headOption
-      tier.fold[OutboundMessageChange](MembershipDeletion(id)) { t =>
-        val num = (obj \ "Membership_Number__c").headOption.map(_.text)
-        MembershipUpdate(Attributes(id, t, num))
+    def orgIdValidation: OutboundMessageParseError \/ Boolean = {
+      for {
+        orgId <- (payload \\ "notifications").getText("OrganizationId")
+        isValid <- if (orgId === organizationId) true.right else OrgIdMatchingError(orgId).left
+      } yield isValid
+    }
+
+    def getSalesforceObjects: OutboundMessageParseError \/ NodeSeq = {
+        val salesforceObjects = (payload \\ "sObject")
+        if (salesforceObjects.isEmpty) {
+          ParsingError(s"No Salesforce Objects were found when parsing the message. \n $payload").left
+        }
+        else salesforceObjects.right
+    }
+
+    def processSalesforceObject(salesforceObject: Node): OutboundMessageChange = {
+      val id = (salesforceObject \ "IdentityID__c").map(_.text).head
+      val tier = (salesforceObject \ "Membership_Tier__c").map(_.text).headOption
+      val num = (salesforceObject \ "Membership_Number__c").headOption.map(_.text)
+      // Match on the Tier to determine the required action
+      tier match {
+        // If Salesforce Contact object has no Tier, we assume the user has expired/cancelled and mark them for deletion
+        case None => MembershipDeletion(id)
+        // If the Salesforce Contact has a Tier, we mark them for an update
+        case Some(tier) => MembershipUpdate(Attributes(id, tier, num))
       }
     }
+
+    orgIdValidation match {
+      case -\/(ParsingError(msg)) => ParsingError(msg).left
+      case -\/(OrgIdMatchingError(orgId)) => OrgIdMatchingError(orgId).left
+      // If we can validate the organization ID, then we get a Right and can attempt to process the notifications.
+      case \/-(_) => getSalesforceObjects.map { _.map(sfObject => processSalesforceObject(sfObject)) }
```

Comment:
not sure what the general wisdom is, but I feel like we could just map the orgidvalidation with getSalesforceObjects, and then flatMap that with the bit you put in braces there, rather than breaking it out with a pattern match and then putting it back together